extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -0,0 +1,127 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as yaml from 'js-yaml';
+import { Book } from './bookTreeItem';
+
+export class BookTreeViewProvider implements vscode.TreeDataProvider<Book> {
+
+	private _onDidChangeTreeData: vscode.EventEmitter<Book | undefined> = new vscode.EventEmitter<Book | undefined>();
+	readonly onDidChangeTreeData: vscode.Event<Book | undefined> = this._onDidChangeTreeData.event;
+	private tocPath: string[];
+
+	constructor(private workspaceRoot: string) {
+		if (workspaceRoot !== '') {
+			this.tocPath = this.getFiles(this.workspaceRoot, []);
+			if (this.tocPath === undefined || this.tocPath.length === 0) {
+				vscode.commands.executeCommand('setContext', 'bookOpened', false);
+			} else {
+				vscode.commands.executeCommand('setContext', 'bookOpened', true);
+			}
+		}
+	}
+
+	private getFiles(dir: string, files_: string[]): string[] {
+		files_ = files_ || [];
+		let files = fs.readdirSync(dir);
+		for (let i in files) {
+			let name = dir + '/' + files[i];
+			if (fs.statSync(name).isDirectory()) {
+				this.getFiles(name, files_);
+			} else if (files[i] === 'toc.yml') {
+				files_.push(name);
+			}
+		}
+		return files_;
+	}
+
+	async openNotebook(resource: vscode.Uri): Promise<void> {
+		let doc = await vscode.workspace.openTextDocument(resource);
+		vscode.window.showTextDocument(doc);
+	}
+
+	getTreeItem(element: Book): vscode.TreeItem {
+		return element;
+	}
+
+	getChildren(element?: Book): Thenable<Book[]> {
+		if (element) {
+			if (element.toc) {
+				return Promise.resolve(this.getSections(element.toc, element.root));
+			} else {
+				return Promise.resolve([]);
+			}
+		} else {
+			return Promise.resolve(this.getBooks());
+		}
+	}
+
+	private getBooks(): Book[] {
+		let books: Book[] = [];
+		for (let i in this.tocPath) {
+			let root = this.tocPath[i].substring(0, this.tocPath[i].lastIndexOf('_data/toc.yml'));
```

Comment:
can you investigate the path library here instead of doing a substring?

Reply:
changed to:
let root = path.dirname(path.dirname(this.tableOfContentsPath[i]));